src/background/services/deduplicator.test.ts

```diff
@@ -0,0 +1,241 @@
+import { Deduplicator } from './deduplicator';
+import { Logger } from '@/shared/logger';
+
+describe('Deduplicator', () => {
+  const deduplicatorService: Deduplicator = new Deduplicator({
+    logger: {
+      debug: jest.fn(),
+    } as unknown as Logger,
+  });
+
+  beforeAll(async (): Promise<void> => {
+    jest.useFakeTimers();
+  });
+
+  beforeEach(() => {
+    jest.runAllTimers();
+  });
+
+  // utility function to create async functions for testing
+  const createAsyncFn = <T>({
+    returnValue,
+    timeout = 0,
+    shouldReject = false,
+  }: {
+    returnValue: T;
+    timeout?: number;
+    shouldReject?: boolean;
+  }) =>
+    jest.fn(
+      async (..._args: unknown[]) =>
+        new Promise((resolve, reject) => {
+          try {
+            if (shouldReject) {
+              reject(new Error('Test error'));
+            }
+            setTimeout(() => {
+              resolve(returnValue);
+            }, timeout);
+          } catch (e) {
+            reject(e);
+          }
+        }),
+    );
+
+  describe('Basic Deduplication', () => {
+    it('should call the original function only once for multiple simultaneous calls', async () => {
+      const returnValue = {
+        access_token: { value: 'value', access: { type: 'incoming' } },
+      };
+      const fn = createAsyncFn({ returnValue });
+      const dedupedFn = deduplicatorService.dedupe(fn);
+      const resultPromises = [dedupedFn(), dedupedFn(), dedupedFn()];
+      jest.runAllTimers();
+
+      const results = await Promise.all(resultPromises);
+      expect(results[0]).toBe(returnValue);
+      expect(results[1]).toBe(returnValue);
+      expect(results[2]).toBe(returnValue);
+      expect(fn).toHaveBeenCalledTimes(1);
+    });
+
+    it('should support different function signatures', async () => {
+      const fn1 = createAsyncFn({
+        returnValue: 'hello',
+        timeout: 100,
+      });
+      Object.defineProperty(fn1, 'name', { value: 'fn1' });
+      // createAsyncFn function returns an anonymous function, which is created using the new Promise constructor.
+      // it does not have a `name` property, so it does NOT have a key for deduplication service
+      const fn2 = createAsyncFn({
+        returnValue: 'world',
+        timeout: 400,
+      });
+      Object.defineProperty(fn2, 'name', { value: 'fn2' });
```

Comment:
The jest mockName thing didn't work?
Either way, let's pass name as an optional param to the create call, and use this logic within there (conditionally).